canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,864 @@
+{
+    if (lastKeyFrameTime != 0)
+    {
+        (*mapPtr)[lastKeyFrameTime / HUNDREDS_OF_NANOS_IN_A_MILLISECOND] = curKeyFrameTime / HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+        auto iter = mapPtr->begin();
+        while (iter != mapPtr->end()) {
+            // clean up map: removing timestamps older than 5 min from now
+            if (iter->first < (duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count() - (300000)))
+            {
+                iter = mapPtr->erase(iter);
+            } else {
+                break;
+            }
+        }
+    }
+    lastKeyFrameTime = curKeyFrameTime;
+}
+
+VOID pushErrorMetrics(Frame frame, CustomData *cusData, double duration)
+{
+    Aws::CloudWatch::Model::MetricDatum metricDatum;
+    Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+    cwRequest.SetNamespace("KinesisVideoSDKCanary");   
+
+    //auto stream_metrics = cusData->kinesisVideoStream->getMetrics();
+    auto stream_metrics_raw = cusData->kinesisVideoStream->getMetrics().getRawMetrics();
+
+    UINT64 newPutFrameErrors = stream_metrics_raw->putFrameErrors - cusData->totalPutFrameErrorCount;
+    cusData->totalPutFrameErrorCount = stream_metrics_raw->putFrameErrors;
+    double putFrameErrorRate = newPutFrameErrors / (double)duration;
+    pushMetric("PutFrameErrorRate", putFrameErrorRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pDimensionPerStream, cwRequest);
+    LOG_DEBUG("PutFrame Error Rate: " << putFrameErrorRate);
+
+    UINT64 newErrorAcks = stream_metrics_raw->errorAcks - cusData->totalErrorAckCount;
+    cusData->totalErrorAckCount = stream_metrics_raw->errorAcks;
+    double errorAckRate = newErrorAcks / (double)duration;
+    pushMetric("ErrorAckRate", errorAckRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pDimensionPerStream, cwRequest);
+    LOG_DEBUG("Error Ack Rate: " << errorAckRate);
+
+    UINT64 totalNumberOfErrors = cusData->totalPutFrameErrorCount + cusData->totalErrorAckCount;
+    pushMetric("TotalNumberOfErrors", totalNumberOfErrors, Aws::CloudWatch::Model::StandardUnit::Count, metricDatum, cusData->pDimensionPerStream, cwRequest);
+    LOG_DEBUG("Total Number of Errors: " << totalNumberOfErrors);
+
+    if (cusData->pCanaryConfig->useAggMetrics)
+    {
+        pushMetric("PutFrameErrorRate", putFrameErrorRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pAggregatedDimension, cwRequest);
+        pushMetric("ErrorAckRate", errorAckRate, Aws::CloudWatch::Model::StandardUnit::Count_Second, metricDatum, cusData->pAggregatedDimension, cwRequest);
+        pushMetric("TotalNumberOfErrors", totalNumberOfErrors, Aws::CloudWatch::Model::StandardUnit::Count, metricDatum, cusData->pAggregatedDimension, cwRequest);
+    }
+
+    // Send metrics to CW
+    cusData->pCWclient->PutMetricDataAsync(cwRequest, onPutMetricDataResponseReceivedHandler);
```

Comment:
Again. Not sure I understand this. Using pushMetric, you are setting cwRequest.AddMetricDatum = metricDatum. This call to `pushMetric` happens for every unique metric, however, your PutMetricDataAsync call is made once? Wouldnt the cwRequest get overwritten? With this call, it seems like you are actually only publishing only aggregated metrics if useAggMetrics is set and totalNumberOfErrors if not. You will need this line after every metric.